### PR TITLE
[WIP] Spacemen are now Forklift Certified. Adds the Rapid Construction Forklifts to the game.

### PR DESCRIPTION
**Pull request by Iamgoofball**
## Early demo video:

https://user-images.githubusercontent.com/4081722/174664585-d4115a2a-9392-4ec7-9081-6a6fa360c767.mp4

## "How do I use this forklift?"

LMB to place. RMB to cancel/refund compatible item. Can't refund anything with contents or that isn't placable from that module.
Scroll wheel to select a new object from the current module.
Shift+Scrollwheel to change module.
Ctrl+Scrollwheel on most modules rotates (counter)clockwise.
Alt+Scrollwheel cycles through additional options, if available.
Middle Mouse also cycles through additional options, if available.

I'll finish the changelog and shit when it's done.

## TODO list:
- [x] Setup forklifts
- [x] Setup forklift modules
- [x] Setup construction holograms
- [x] Floors module
- [x] Walls module
- [x] Furniture module
- [x] Airlocks module
- [x] Departmental Machinery module
- [x] Shuttle module
- [x] Define departmental forklifts
- [x] Put keys in Head of Staff lockers
- [ ] Put keys in hands of highest hour count staffers in departments
- [ ] Find way to get more forklifts, cargo order maybe?
- [x] Put sprites in so it's not just the janicart

### CODE REVIEW

**Comment by Addust** [comment]
Merge now

**Comment by Tennessee116** [comment]
Dear God.

**Comment by CursedBirb** [comment]
Link to design doc for anyone interested
https://hackmd.io/@tgstation/rJjgKJEYc

**Comment by dalmationer** [comment]
What are the plans for cyborgs? Are they to keep the RCDs they have currently or will they move over to this system?
I kind of like the idea of cyborgs havin this sorta thing built in. might be cool!

**Comment by dalmationer** [comment]
Also I'm happy to cook up some sprites for this if need be. I've been working on some digitigrade jumpsuits, shoes etc. and an actual sprite for the mechanical wheelchair, but if yous need somethin for this i could cook it up.
I think it should be some kinda big chunky construction vehicle lookin kinda thing that you sit inside.

**Comment by MrDoomBringer** [comment]
From my understanding of the design doc, the intention was to have a 'control tool' that acted as an RCD that you took from the forklift and was usable within a few tiles of the forklift, rather than what I see in the video. 

This interacts with features like multiple users and the auto-follow mode of the forklift. How do you plan to realize multiple users, the auto-follow-mode, etc. in this implementation?

**Comment by Iamgoofball** [comment]
follow mode isn't feasible in a performant and useful way that isn't outclassed on a technical and gameplay level by pulling 

multi-user support is built in

**Comment by LemonInTheDark** [comment]
Needs a ui

File: code/game/atoms.dm [hunk 1]
```diff
@@ -1952,6 +1952,8 @@
 	return TRUE
 
 /atom/MouseEntered(location, control, params)
+	if(flags_1 & INITIALIZED_1)
+		usr.MouseEnteredOn(src, location, control, params)
```
Comment (Mothblocks) [inline code/game/atoms.dm:1956]: This is going to be bad for performance, I separated mouse enters into their own very very optimized subsystem for a reason. CC @LemonInTheDark

**Comment by Shadowflame909** [comment]
Bombs?

We have the power to rebuild!

File: code/game/atoms.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/game/atoms.dm:1956]: We're already eating the perf hit by having on mouse entered defined at the atom level. I'm concerned that switching to.the subsystem would break ECS design and actually be slower and less responsive to players.

**Comment by ZerVexon** [comment]
So not only is it longer ranged than the RCD, its faster?

Looks like a really, really good tool to just grief everyone by deleting floors everywhere. And since its an engineering cart, that means Engineers have access to it, and Engineers are probably the best equipped for dealing with space, since it's literally their job. 

Also worse than jaws or RCD at telling access to go fuck itself since it can apparently delete doors, windows, and walls. Oh, and you can't be shoved or knocked down either, since you're buckled.

This is the stuff of nightmares.

File: code/game/atoms.dm (hunk 1 above)
Comment (Mothblocks) [reply code/game/atoms.dm:1956]: We are taking *a* perf hit, but by adding an extra proc to this, especially during verb time, you are going to notice a *significant* downgrade in performance, the *whole reason* it was moved over to a subsystem



If this is for the placement silhouette, the cheapest way to do this is most likely to create a full screen UI when you are in the forklift, and use the mouse position to calculate what tile they're hovering over.

**Comment by Iamgoofball** [comment]
@ZerVexon wow it's almost like the video is a WIP or something

**Comment by ZerVexon** [comment]
> @ZerVexon wow it's almost like the video is a WIP or something

So, by it's very nature, you're asking for feedback and, by proxy, any potential issues?

**Comment by MrDoomBringer** [comment]
>for obvious reasons, the final product will not let you instantly deconstruct things

It would be exceedingly helpful if you actually said in the PR what parts of the design doc you're ignoring and what parts you just haven't finished yet, as right now the distinction is unclear. If you're unwilling to do this, don't open a PR. 

What's obvious to you isn't necessarily obvious to the casual observer.

File: code/modules/vehicles/forklift/forklift.dm [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	UnregisterSignal(M, list(COMSIG_MOUSE_SCROLL_ON, COMSIG_MOB_CLICKON, COMSIG_MOUSE_ENTERED_ON, COMSIG_MOB_SAY))
+	qdel(selected_modules[M])
+	..()
+
+/obj/vehicle/ridden/forklift/key_inserted()
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/vehicle/ridden/forklift/process(delta_time)
+	if(COOLDOWN_FINISHED(src, build_cooldown)) // Build a hologram!
+		for(var/obj/structure/building_hologram/hologram in holograms)
```
Comment (Kylerace) [inline code/modules/vehicles/forklift/forklift.dm:79]: you never remove holograms from this list as far as i can tell

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/game/atoms.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/game/atoms.dm:1956]: I told Goof before that he needs to make this a contextual UI since he wants to be able to check a mouse enter for every tile they hover

Comment (LemonInTheDark) [inline code/game/atoms.dm:1956]: The proper pattern for this is a fullscreen object to act as a catcher. see `/atom/movable/screen/fullscreen/scope`
This ain't acceptable in any way, this proc is called for every object a client moves their mouse over in the space of a tick, 10 is maybe the worst case scenario

This code needs to be as lightweight as is possible.

**Comment by kopoba** [comment]
If you slliped on this thing you still can place stuff (hologram/items)walking around this car dunno if this intended(looks like no). And you stuck with this hologram projector thingy for the rest of your miserable life. Moving from it doesn't remove this holo placing/switching ability and you can proceed placing them to the limit. Also you cant use hands with this hologram stucked to you. Mine forklift got stolen by stupid catgirl after i sliped. I try to sleep, reconnect, buckle/unblucle to other forklift to fix this. So i just suicide without option to get my forklift back and unbug ability to use hands =(
round id 185702 if this matter
image

**Comment by Pinkufeck** [comment]
Going out of it doesn't delete the construction hologram of what you are planning to build. Impossible to deconstruct/refund anything because it's always on cooldown. Adding a description in game for the controls might be wise too cause right now it's not very intuitive. Interacting with your inventory while on it will just make it build under you. Can't fix floors, can't fix walls. And can't take out mats from it so unable to even take iron out of it to repair.
Basically a worse RCD that can't even fix breaches. Very situational and only "useful" if you are trying to decorate some place by adding chairs and such. Probably won't be used for that either because building chairs for twice their regular costs doesn't outweight the speed at which it builds them.

**Comment by kopoba** [comment]
> Going out of it doesn't delete the construction hologram of what you are planning to build. Impossible to deconstruct/refund anything because it's always on cooldown. Adding a description in game for the controls might be wise too cause right now it's not very intuitive. Interacting with your inventory while on it will just make it build under you. Can't fix floors, can't fix walls. And can't take out mats from it so unable to even take iron out of it to repair. Basically a worse RCD that can't even fix breaches. Very situational and only "useful" if you are trying to decorate some place by adding chairs and such. Probably won't be used for that either because building chairs for twice their regular costs doesn't outweight the speed at which it builds them.

im kinda sure you can place walls well there was an option for that when i was checking options that forklift can do its not only decoration.

**Comment by Shadowflame909** [comment]
Wheres the forward sprite..your invisible if you go downwards

**Comment by Pickle-Coding** [comment]
It would be nice if there was a timer to unbuckle from the forklift, and if there was a significant movement penalty if you are dragging the forklift instead of driving it. I don't like how any tider can instantly initiate a buckle war with forklifts if you need to use them for something.

**Comment by Pinkufeck** [comment]
Also what's the material cost multiplier compared to a RCD ?

**Comment by Mazur907** [comment]
it would be fun to emag it and turn it into a maniacal killing machine

**Comment by Mothblocks** [comment]
still conflicting and still has the untenable MouseEntered base impl

**Comment by Mothblocks** [comment]
We have too many open PRs and this has received no updates in a significant amount of time as well as having a huge amount of conflicts

**Comment by deathrobotpunch** [comment]
> Middle Mouse also cycles through additional options, if available

please I beg of you add these to the hotkey settings because my middle mouse doesn’t work

**Comment by optimumtact** [comment]
make a new PR if you intend to work on this.

**Comment by optimumtact** [comment]
@ZephyrTFA don't reopen very old PR's, make them come to the top of the queue with a new PR instead.

**Comment by Iamgoofball** [comment]
no worries wasn't sure what protocol was

File: code/game/atoms.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/game/atoms.dm]: migrated it to a signal that's off of on_mouse_enter/SSmouse_entered, still laggy as hell but gets the job done-ish

https://github.com/tgstation/tgstation/assets/4081722/b9aa9aa0-6e1b-4af1-8ac6-cc9eb69c21e4

Comment (Iamgoofball) [reply code/game/atoms.dm]: i actually had a bug, this is now lightning fast, fuck yes great work on the subsystem

https://github.com/tgstation/tgstation/assets/4081722/c96e8a94-b312-43b7-a516-59d1c5ede5b4
